### PR TITLE
handle conditionals in duplicate module checks

### LINKED ISSUE #4629: cabal check doesn't account for conditionals
In [`binary-serialise-cbor`]() we are trying to provide `exposed-modules` for GHC 7.8, where `reexported-modules` is not supported. For instance,
```
library
  default-language:  Haskell2010
  hs-source-dirs:      src
  if impl(ghc >= 7.9)
    reexported-modules:
      Codec.Serialise            as Data.Binary.Serialise.CBOR,
      Codec.Serialise.Class      as Data.Binary.Serialise.CBOR.Class,
      Codec.Serialise.Decoding   as Data.Binary.Serialise.CBOR.Decoding,
      Codec.Serialise.Encoding   as Data.Binary.Serialise.CBOR.Encoding,
      Codec.CBOR.FlatTerm        as Data.Binary.Serialise.CBOR.FlatTerm,
      Codec.Serialise.IO         as Data.Binary.Serialise.CBOR.IO,
      Codec.CBOR.Magic           as Data.Binary.Serialise.CBOR.Magic,
      Codec.CBOR.Pretty          as Data.Binary.Serialise.CBOR.Pretty,
      Codec.Serialise.Properties as Data.Binary.Serialise.CBOR.Properties,
      Codec.CBOR.Write           as Data.Binary.Serialise.CBOR.Write,
      Codec.CBOR.Term            as Data.Binary.Serialise.CBOR.Term,
      Codec.Serialise.Tutorial   as Data.Binary.Serialise.CBOR.Tutorial
    exposed-modules: Data.Binary.Serialise.CBOR.Read
  else
    -- GHC 7.8 doesn't support module re-exports
    exposed-modules:
      Data.Binary.Serialise.CBOR,
      Data.Binary.Serialise.CBOR.Class,
      Data.Binary.Serialise.CBOR.Decoding,
      Data.Binary.Serialise.CBOR.Encoding,
      Data.Binary.Serialise.CBOR.FlatTerm,
      Data.Binary.Serialise.CBOR.IO,
      Data.Binary.Serialise.CBOR.Magic,
      Data.Binary.Serialise.CBOR.Pretty,
      Data.Binary.Serialise.CBOR.Properties,
      Data.Binary.Serialise.CBOR.Read,
      Data.Binary.Serialise.CBOR.Write,
      Data.Binary.Serialise.CBOR.Term

  build-depends:
    base < 5,
    bytestring < 1.0,
    cborg     == 0.1.*,
    serialise == 0.1.*

```
Unfortunately, `cabal check` fails on this with,
```
$ cabal check
The package will not build sanely due to these errors:
* Duplicate modules in library: Data.Binary.Serialise.CBOR,
Data.Binary.Serialise.CBOR.Class, Data.Binary.Serialise.CBOR.Decoding,
Data.Binary.Serialise.CBOR.Encoding, Data.Binary.Serialise.CBOR.FlatTerm,
Data.Binary.Serialise.CBOR.IO, Data.Binary.Serialise.CBOR.Magic,
Data.Binary.Serialise.CBOR.Pretty, Data.Binary.Serialise.CBOR.Properties,
Data.Binary.Serialise.CBOR.Read, Data.Binary.Serialise.CBOR.Term,
Data.Binary.Serialise.CBOR.Write

Hackage would reject this package.
```

It seems that Cabal doesn't account for the conditional when computing this check.

### PR DESCRIPTION
**Pull request by gbaz**
Resolves #4629

This is a less invasive approach than my last pass. It only factors out the duplicate module checks without touching anything else. This is a better "rough cut" approximation but still not perfect. In particular, It won't generate duplicate module warnings from a module existing on both sides of an `if/else` conditional -- however, if there are distinct conditional blocks which logically are exclusive (i.e. one is a check on a positive condition and the other a check on its negation) it won't catch that by actually model-checking the various possible outcomes.

### CODE REVIEW

**Review by emilypi — APPROVED** [review]
This looks good @gbaz. Clean solution

**Comment by emilypi** [comment]
Tho we do need a changelog.

File: Cabal/src/Distribution/PackageDescription/Check.hs [hunk 1]
```diff
@@ -2158,6 +2129,29 @@ checkSetupVersions pkg =
...
         ++ "version."
 
+checkDuplicateModules :: GenericPackageDescription -> [PackageCheck]
+checkDuplicateModules pkg =
+       concatMap checkLib   (maybe id (:) (condLibrary pkg) . map snd $ condSubLibraries pkg)
+    ++ concatMap checkExe   (map snd $ condExecutables pkg)
+    ++ concatMap checkTest  (map snd $ condTestSuites  pkg)
+    ++ concatMap checkBench (map snd $ condBenchmarks  pkg)
+  where
+    -- the duplicate modules check is has not been thoroughly vetted for backpack
```
[the remark this answers is not in the document]
Comment (gbaz) [reply Cabal/src/Distribution/PackageDescription/Check.hs:2139]: this was me migrating the old comment -- the current logic is no worse than before, but i didn't want to lose a potential warning here :-)

**Review by Mikolaj — APPROVED** [review]
Great job, this produces much better approximation, while staying linear and improving code quality.

**Pull request merged**